### PR TITLE
feat(routes): expose findRoute and param validator

### PR DESCRIPTION
**Pull request by sf3ris**
### Description
Quick draft of implementation to solve https://github.com/fastify/fastify/issues/5215
In this PR I wanted to expose an abstraction to retrieve a route for doing further checks.
Furthermore, I was doubtful about creating a standalone plugin for a simple check like this but open for any discussion.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/route.js [hunk 1]
```diff
@@ -168,11 +170,31 @@ function buildRouting (options) {
...
+    }
+
+    const schemaParams = new Set(Object.keys(route.store.schema.params))
+    const params = new Set(Object.keys(route.params))
+    if (schemaParams.size !== params.size) {
+      return false
+    }
+
+    return [...schemaParams].every(param => params.has(param))
   }
```
Comment (metcoder95) [inline lib/route.js:198]: Overall, I'm +1 on exposing `findRoute` but not in sync with having this validation within core. The scope for the feature seems really specific, that's why the suggestion of having it within a community plugin.

File: test/findRoute.test.js
Comment (metcoder95) [inline test/findRoute.test.js:1]: Can you add some tests with this being used from a plugin?

**Review by mcollina — CHANGES_REQUESTED** [review]
We support different kinds of validators. I don't think we should tie with a specific one and having the validations built in. It's likely better to have them in a separate plugin.

**Comment by sf3ris** [comment]
@mcollina @metcoder95 Makes sense to me, just a bit doubtful about the scope of the plugin.
What do you think about just exposing the `findRoute` as you suggested and leaving the plugin creation\implementation to later discussions?

**Comment by metcoder95** [comment]
SGTM 👍

File: test/findRoute.test.js
Comment (sf3ris) [reply test/findRoute.test.js:1]: Added here https://github.com/fastify/fastify/pull/5230/commits/b3ad236cd2c95b905e9e072884ce7f11e3ecf4e6, LMK whether you think is enough or you have other types of tests in mind 🙏

File: fastify.js [hunk 2]
```diff
@@ -285,6 +285,9 @@ function fastify (options) {
     hasRoute: function _route (options) {
       return router.hasRoute.call(this, { options })
     },
+    findRoute: function _route (options) {
```
Comment (metcoder95) [inline fastify.js:288]:
```suggestion
    findRoute: function _findRoute (options) {
```

**Review by metcoder95 — COMMENTED** [review]
Nice! LGTM, small `nit`

File: types/instance.d.ts [hunk 3]
```diff
@@ -219,6 +219,12 @@ export interface FastifyInstance<
     SchemaCompiler extends FastifySchema = FastifySchema,
   >(opts: Pick<RouteOptions<RawServer, RawRequest, RawReply, RouteGeneric, ContextConfig, SchemaCompiler, TypeProvider>, 'method' | 'url' | 'constraints'>): boolean;
 
+  findRoute<
+    RouteGeneric extends RouteGenericInterface = RouteGenericInterface,
+    ContextConfig = ContextConfigDefault,
+    SchemaCompiler extends FastifySchema = FastifySchema,
+  >(opts: Pick<RouteOptions<RawServer, RawRequest, RawReply, RouteGeneric, ContextConfig, SchemaCompiler, TypeProvider>, 'method' | 'url' | 'constraints'>): FindResult<FindMyWayVersion<RawServer>>;
```
Comment (mcollina) [inline types/instance.d.ts:226]: Could you add a test for this type? We use tsd.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by metcoder95 — APPROVED** [review]

**Pull request merged**